internal/provider/scheduler/farmer_bot.go

```diff
@@ -0,0 +1,169 @@
+	PublicConfig *bool    `json:"public_config,omitempty"`
+	PublicIPs    *uint32  `json:"public_ips"`
+	Certified    *bool    `json:"certified,omitempty"`
+}
+
+type Params struct {
+	Key   string      `json:"key"`
+	Value interface{} `json:"value"`
+}
+
+func (s *Scheduler) hasFarmerBot(ctx context.Context, farmID uint32) bool {
+	_, err := s.getFarmInfo(farmID)
+	if err != nil {
+		return false
+	}
+	args := []Args{}
+	params := []Params{}
+	data := s.generateFarmerBotAction(farmID, args, params, FarmerBotVersionAction)
+	var output FarmerBotAction
+	dst := s.farms[farmID].farmerTwinID
+	err = s.rmbClient.Call(ctx, dst, FarmerBotRMBFunction, data, &output)
+	if err != nil {
+		log.Printf("error pinging farmerbot %+v", err)
+	}
+
+	return err == nil
+}
+
+func (n *Scheduler) farmerBotSchedule(ctx context.Context, r *Request) (uint32, error) {
+	info, err := n.getFarmInfo(r.FarmId)
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to get farm %d info", r.FarmId)
+	}
+	params := generateFarmerBotParams(r)
+	args := generateFarmerBotArgs(r)
+	data := n.generateFarmerBotAction(info.farmerTwinID, args, params, FarmerBotFindNodeAction)
+	output := FarmerBotAction{}
+
+	err = n.rmbClient.Call(ctx, info.farmerTwinID, FarmerBotRMBFunction, data, &output)
+	if err != nil {
+		return 0, err
+	}
+	if len(output.Result.Params) < 1 {
+		return 0, errors.New("can not find a node to deploy on")
+	}
+	nodeId, err := strconv.ParseUint(output.Result.Params[0].Value.(string), 10, 32)
+	if err != nil {
+		return 0, err
+	}
+	log.Printf("got a node with id %d", nodeId)
+	return uint32(nodeId), nil
+}
+
+func generateFarmerBotArgs(r *Request) []Args {
+	return []Args{}
+}
+func generateFarmerBotParams(r *Request) []Params {
+	params := []Params{}
+	if r.Capacity.HRU != 0 {
+		params = append(params, Params{Key: "required_hru", Value: r.Capacity.HRU})
+	}
+
+	if r.Capacity.SRU != 0 {
+		params = append(params, Params{Key: "required_sru", Value: r.Capacity.SRU})
+	}
+
+	if r.Capacity.MRU != 0 {
+		params = append(params, Params{Key: "required_mru", Value: r.Capacity.MRU})
+	}
+
+	if r.Capacity.CRU != 0 {
+		params = append(params, Params{Key: "required_cru", Value: r.Capacity.CRU})
+	}
+
+	if len(r.NodeExclude) != 0 {
+		value := strings.Trim(strings.Join(strings.Fields(fmt.Sprint(r.NodeExclude)), ","), "")
+		params = append(params, Params{Key: "node_exclude", Value: value})
+	}
+
+	if r.Dedicated {
+		params = append(params, Params{Key: "dedicated", Value: r.Dedicated})
+	}
+
+	if r.PublicConfig {
+		params = append(params, Params{Key: "public_config", Value: r.PublicConfig})
+	}
+
+	if r.PublicIpsCount > 0 {
+		params = append(params, Params{Key: "public_ips", Value: r.PublicIpsCount})
+	}
+
+	if r.Certified {
+		params = append(params, Params{Key: "certified", Value: r.Certified})
+	}
+	return params
+}
+
+func (s *Scheduler) generateFarmerBotAction(farmerTwinID uint32, args []Args, params []Params, action string) FarmerBotAction {
```

Comment:
Why does this live on the scheduler receiver? if for source twin, please make it a parameter instead